core/domain/email_manager_test.py

```diff
@@ -2015,6 +2232,42 @@ def test_that_correct_email_is_sent(self):
             sent_email_model.intent,
             feconf.BULK_EMAIL_INTENT_MARKETING)
 
+    def test_that_email_not_sent_if_cleaned_not_equal_html(self):
+        email_subject = 'Dummy Email Subject'
+        email_html_body = 'Dummy email body.<td>'
+
+        logged_errors = []
+
+        def _log_error_for_tests(error_message):
+            """Appends the error message to the logged errors list."""
+            logged_errors.append(error_message)
+
+        log_new_error_counter = test_utils.CallCounter(_log_error_for_tests)
+        log_new_error_ctx = self.swap(
+            email_manager, 'log_new_error', log_new_error_counter)
+        with self.can_send_emails_ctx, log_new_error_ctx:
+            self.assertEqual(log_new_error_counter.times_called, 0)
+            # pylint: disable=protected-access
+            email_manager._send_bulk_mail(
```

Comment:
Avoid testing private api. Try testing it through some public function which uses the API. In general follow rule 9 in the guideline [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

Reply:
Done

Reply:
@bansalnitish @DubeySandeep 
Sorry, I misinterpreted and could not understand what you actually meant. I want to correct the mistake now. I need a few suggestions.

I missed that point while writing the tests. I tried to find a stub for the _send_bulk_mail that provides access to the private function, however, could not find it.

I was looking through other related files and found that send_bulk_mail was implemented as a non-private function in core/platform/email/mailgun_email_services_test.py.

What do you suggest? Should I go ahead with creating a stub for the function in emai_manager_test.py file itself or convert _send_bulk_mail to send_bulk_mail?

Looking forward to your suggestions.

Reply:
@siddhantkhandelwal,
 > I was looking through other related files and found that send_bulk_mail was implemented as a non-private function in core/platform/email/mailgun_email_services_test.py.

I cannot find such method in `mailgun_email_services_test` can you add the line number?

Regarding guidelines for writing test related to private methods, I think you'll find [this wiki page](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests) useful.  (see 9th bullet)

Reply:
@DubeySandeep 

In the file core/platform/email/gae_email_services.py on line number 85, send_bulk_mail is defined as a non-private function. We can do this similarly for the function in email_manage.py

On the other hand, I can create a stub send_bulk_mail that handles the private function _send_bulk_mail in email_manager.py

Looking forward to your suggestions!
Thanks!

Reply:
Hi @siddhantkhandelwal, I can see that `_send_bulk_email` is called in `send_user_query_email` [here](https://github.com/oppia/oppia/blob/develop/core/domain/email_manager.py#L954). Can we try to test through this?

Reply:
@bansalnitish 
Yes, we can do this. 
Just one small thing though, I am not clear with the structure here.
Since I am testing _send_bulk_mail function, is it not that I am supposed to use only that function or it's stub. Using some other function that in turn uses the testee function kind of defeats the ideology, doesn't it?
I am unable to understand this. Can you clear this to me?
Thanks!

Reply:
Since we are testing a private api, we should try testing it through some public interface. If you test it through the public interface, you can test each behaviour for it by using the same cases as you were using when calling it directly. So, it fulfils the ideology as well as avoids direct access to private api.

Reply:
Do let me know if you have you have any other doubts on this.

Reply:
Hey @bansalnitish,
Oh, I see. I will make changes accordingly.
Thanks for the help!